File: components/theme_manager/theme_manager.py
import mesop as me
import typing

@me.web_component(path="./theme_manager.js")
def theme_manager(
    *,
    theme: str,
    on_theme_load: typing.Callable[[me.WebEvent], None],
):
    return me.insert_web_component(
        name="theme-manager",
        properties={
            "theme": theme,
        },
        events={
            "themeLoaded": on_theme_load,
        },
    )

File: components/theme_manager/theme_manager.js
import { LitElement } from 'https://esm.sh/lit';

class ThemeManager extends LitElement {
  static get properties() {
    return {
      theme: { type: String },
      themeLoaded: { type: String },
    };
  }

  constructor() {
    super();
    this.theme = '';
  }

  connectedCallback() {
    super.connectedCallback();
    const storedTheme = localStorage.getItem('theme');
    if (storedTheme) {
      this.dispatchEvent(new MesopEvent(this.themeLoaded, { theme: storedTheme }));
    }
  }

  updated(changedProperties) {
    if (changedProperties.has('theme') && this.theme) {
      localStorage.setItem('theme', this.theme);
    }
  }

  render() {
    // This component is non-visual
    return '';
  }
}

customElements.define('theme-manager', ThemeManager);